[Feat] List, Select 컴포넌트 생성

aken-you opened the pull request:
# 관련 이슈 번호

#107 

# 설명

## List 컴포넌트 구성

List 컴포넌트는 크게 2개의 컴포넌트로 구성되어 있어요.

1. List (= ListMain)
2. Item

### 1. List

항목들을 감싸는 컴포넌트입니다.

```typescript
type ListProps = HTMLAttributes<HTMLUListElement>;

const ListMain = ({ children, ...props }: ListProps) => {
  return (
    <ul className="flex w-full flex-col justify-center" {...props}>
      {children}
    </ul>
  );
};
```

### 2. Item

- ref: 나중에 키보드 접근성 로직을 구현하기 위해, 미리 forwardRef로 감쌌습니다.
- props
  - disabled: 항목 비활성화 여부
  - additionalClassName: `Select.Option`에서 isSelected에 따라 텍스트 색상을 다르게 하기 위해, props로 추가적인 class를 받습니다.

```typescript
interface ItemProps extends LiHTMLAttributes<HTMLLIElement> {
  disabled?: boolean;
  additionalClassName?: string;
}

export const Item = forwardRef<HTMLLIElement, ItemProps>(
  (
    { children, disabled = false, onClick, additionalClassName, ...props },
    ref,
  ) => {
    const baseStyles =
      "flex h-[3.625rem] w-full min-h-12 shrink-0 justify-center items-center gap-[.625rem] px-[.625rem] py-[1.125rem] rounded-2xl title-2";

    const { base, active, hover, focus } = listItemStyles;
    const ableStyles = `${base} ${active} ${hover} ${focus}`;
    const disabledStyles = listItemStyles.disabled;

    return (
      <li
        ref={ref}
        className={`${baseStyles} ${disabled ? disabledStyles : ableStyles} ${additionalClassName}`}
        aria-disabled={disabled}
        onClick={onClick}
        {...props}
      >
        {children}
      </li>
    );
  },
);
```

## List 사용법

```javascript
<List>
        <List.Item>Item 1</List.Item>
        <List.Item>Item 2</List.Item>
        <List.Item>Item 3</List.Item>
</List>
```


##  Select 컴포넌트 시나리오

1. Select 컴포넌트를 렌더링하는 Trigger 요소를 클릭한다.
2. Select 컴포넌트가 모달 또는 바텀 시트 형식으로 렌더링 된다.
3. 리스트에 있는 항목들 중 하나를 클릭한다.
4. Select 컴포넌트는 unmount된다.

## Select 컴포넌트 구성

Select 컴포넌트는 크게 4가지 컴포넌트로 구성되어 있어요.

1. Select
2. SelectBottomSheet
3. OptionList
4. Option

추후에 SelectModal이 추가될 예정입니다.


### 1. Select

(SelectMain이 곧 Select이라고 생각해주시면 됩니다.)

SelectContext.Provider로 감싸져 있고, isOpen과 onClose를 props로 받습니다.
- isOpen: Select 렌더링 여부
- onClose: Select을 닫는 핸들러

```javascript
const SelectMain = ({
  isOpen,
  onClose,
  children,
}: SelectContextType & {
  children: ReactNode;
}) => {
  return (
    <SelectContext.Provider value={{ isOpen, onClose }}>
      {children}
    </SelectContext.Provider>
  );
}
```

#### SelectContext

`SelectContext`에는 `Select` 컴포넌트에서 받은 isOpen과 onClose를 관리합니다.
그리고 context에 있는 값을 사용하기 위해 `useSelectContext` 훅을 호출해야 합니다.
```javascript
export interface SelectContextType {
  isOpen: boolean;
  onClose: () => void;
}

export const SelectContext = createContext<null | SelectContextType>(null);

export const useSelectContext = () => {
  const context = useContext(SelectContext);

  if (!context) {
    throw new Error("useSelectContext must be used within a SelectProvider");
  }

  return context;
};
```

### 2. SelectBottomSheet

`SelectContext`에 있는 isOpen와 onClose를 받아와서, `Sheet` 컴포넌트에 넣어줍니다.

```javascript
const SelectBottomSheet = ({ children }: SelectBottomSheetProps) => {
  const { isOpen, onClose } = useSelectContext();

  return (
    <Sheet isOpen={isOpen} onClose={onClose} detent="content-height">
      <Sheet.Container>
        <Sheet.Header />
        <Sheet.Content>{children}</Sheet.Content>
      </Sheet.Container>
      <Sheet.Backdrop />
    </Sheet>
  );
};
```

### 3. OptionList

앞에서 만든 List 컴포넌트를 이용하였습니다.
2번 SelectBottomSheet의 children을 List 컴포넌트로 감싸줘도 되겠지만,
나중에 Option 컴포넌트를 키보드로 접근하기 위한 로직이 이 컴포넌트에 들어갈 것 같아서 따로 만들어놨습니다.

```javascript
const OptionList = ({
  children,
  ...props
}: HTMLAttributes<HTMLUListElement>) => {
  return <List {...props}>{children}</List>;
};
```

### 4. Option

- List.Item 컴포넌트를 기반으로 합니다
- props
  - isSelected: 선택됐는지 여부
  - disabled: 비활성화 여부
- handleClick: 옵션을 선택하고 나면, Select 컴포넌트가 닫힙니다.
  - SelectContext의 onClose를 이용해, Select 컴포넌트를 unmount합니다.

```javascript
interface OptionProps extends LiHTMLAttributes<HTMLLIElement> {
  isSelected?: boolean;
  disabled?: boolean;
}

const Option = ({
  children,
  isSelected = false,
  disabled = false,
  onClick,
  ...props
}: OptionProps) => {
  const { onClose } = useSelectContext();

  const handleClick = (e: React.MouseEvent<HTMLLIElement>) => {
    onClick?.(e);
    onClose();
  };

  const textColorStyles = disabled
    ? "text-grey-300"
    : isSelected
      ? "text-tangerine-500"
      : "text-grey-700";

  return (
    <List.Item
      onClick={handleClick}
      additionalClassName={textColorStyles}
      {...props}
    >
      {children}
    </List.Item>
  );
};
```

## Select 사용법

```javascript
const Component = () => {
  // select 열림 여부
  const [isOpen, setIsOpen] = useState<boolean>(false);

  // 리스트에서 선택된 항목의 value
  const [selectedValue, setSelectedValue] = useState<number | null>(null);

  // optionList의 타입은 아무거나 해도 됩니다. 저는 아래와 같은 기준으로 optionsList 타입을 설정했습니다.
  // value: 옵션을 선택했을 때 컴포넌트가 내부적으로 사용하는 값 (내부 로직에서만 사용)
  // name: value 대신 화면에 표시되는 텍스트 (ui에 표시)
  const optionList: { value: number; name: string }[] = [ ... ]

  // 트리거 버튼에 표시할 선택된 항목의 텍스트
  const selectedName = optionList.find(
    (option) => option.value === selectedValue,
  )?.name;

  return (
    <>
      {/* select 컴포넌트를 여는 trigger 컴포넌트 */}
      <Trigger
        // ...
        value={selectedName ?? ""}
        onClick={() => setIsOpen(true)}
      />

      <Select isOpen={isOpen} onClose={() => setIsOpen(false)}>
        <Select.BottomSheet>
          <Select.OptionList>
            {optionList.map(({ value, name }) => {
              return (
                <Select.Option
                  key={value}
                  value={value}
                  isSelected={value === selectedValue}
                  onClick={() => setSelectedValue(value)}
                >
                  {name}
                </Select.Option>
              );
            })}
          </Select.OptionList>
        </Select.BottomSheet>
      </Select>
    </>
  );
};

```

aken-you committed:
feat[#107] : List 컴포넌트 생성

aken-you committed:
test[#107] : List 컴포넌트 스토리북 작성

aken-you committed:
feat[#107] : Select 컴포넌트 생성

- Option: 하나의 옵션
- OptionList: 옵션 목록들을 감싸는 컴포넌트
- SelectContext: isOpen, onClose를 저장하는 context
- SelectBottomSheet: OptionList를 감싸는 바텀 시트
- Select: 메인 컴포넌트

aken-you committed:
test[#107] : Select 컴포넌트 스토리북 작성

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/b41e9545a31e5f4f858286a5e6353e39/shared-input.test.js
PASS browser: chromium ../../../../../tmp/b41e9545a31e5f4f858286a5e6353e39/shared-textarea.test.js
PASS browser: chromium ../../../../../tmp/b41e9545a31e5f4f858286a5e6353e39/shared-button.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        4.769 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)

aken-you committed:
refactor[#107] : List의 Item 스타일 객체 네이밍 변경

aken-you committed:
refactor[#107] : List의 Item props에서 isSelected 삭제, additionalClassName 추가

- List에 isSelected가 있는 것보다, Select에서 isSelected 값을 받는 것이 더 자연스럽다고 판단하여 삭제
- additionalClassName을 통해 추가적인 스타일링을 가능하게 함

aken-you committed:
test[#107] : List의 Item 스토리북 작성

aken-you committed:
refactor[#107] : List의 Item 수정에 따라 Select.Option 컴포넌트 수정

aken-you committed:
refactor[#107] : Select의 Option에서 아직까지 쓰지 않는 ref 제거

aken-you committed:
refactor[#107] : List의 role 속성 제거

aken-you committed:
refactor[#107] : List 컴포넌트 export 수정

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/516cc9bb8ba924bbb6b466493d031f90/shared-textarea.test.js
PASS browser: chromium ../../../../../tmp/516cc9bb8ba924bbb6b466493d031f90/shared-input.test.js
PASS browser: chromium ../../../../../tmp/516cc9bb8ba924bbb6b466493d031f90/shared-button.test.js

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        5.565 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)

yonghyeun commented on src/shared/ui/list/List.tsx:
```diff
@@ -0,0 +1,49 @@
+  );
+};
+
+const List = Object.assign(ListMain, {
```
오호 이런식으로 사용하는 방법도 있군요 

```tsx
<List.List>
 <List.Item>
  ...
```

이런식으로 쓸 때 보다 더 깔끔해보이고 멋진 거 같습니다

yonghyeun commented on src/shared/ui/select/Select.stories.tsx:
```diff
@@ -0,0 +1,101 @@
+          onClick={() => setIsOpen(true)}
+        />
+
+        <Select isOpen={isOpen} onClose={() => setIsOpen(false)}>
```
Select 가 State를 내려주는 Context Provider 라는 뉘앙스가 풍기는건 어떨까요 ?

aken-you replied:
State를 내려주는 Context Provider는 `SelectContext.Provider`라고 생각해요.
그리고 `Select`는 `isOpen`과 `onClose`를 `SelectContext.Provider`에 전달해주는 컴포넌트라고 생각하여 `Select` 네이밍이 적절하다고 생각합니다!

yonghyeun reviewed (commented):
정말 깔끔하고 멋진 코드 잘 봤습니다. 

상위에서 상태와 핸들러를 받아 `Select` 컨텍스트를 이용해 하위로 내려주는 아이디어는 정말 멋진 거 같습니다 !! 

스토리북에서 이것 저것 만지다 보니 몇 가지 현상이 발견되어 이야기 드려 봅니다. 

![image](https://github.com/user-attachments/assets/a760d083-d479-4178-a46c-c2a8a1e911ff)

바텀시트의 경우엔 활성화 되어 있을 때 주변이 클릭 가능하여 두 개 이상의 바텀 시트가 동시에 올라오는 현상이 발견 되었습니다. 

바텀 시트 외부가 클릭 되었을 때 open 상태를 `close` 로 바꿀 수 있는 기능이 추가되어야 할 거 같습니다. 혹은 옵선에셔 어떤 값을 내릴 수 있을런지요 

---

+ `Input` 컴포넌트를 만들 때엔 `onClick` 이벤트를 염두하지 않고 만들었습니다.

현재 `Input` 컴포넌트에선 모든 이벤트 핸들러들을 `input` 태그에게만 건내주고 있는데 이로 인해서 

![image](https://github.com/user-attachments/assets/7bea41f4-0f3d-4b0d-8517-ae8e4ae932e8)

`onClick` 이벤트가 저 조그만 `input` 영역을 클릭 할 때에만 트리거가 되더군요 

그래서 하위로 내리는 역삼각형 아이콘을 클릭 해도 바텀 시트가 트리거 되지 않습니다.

이에 `Input` 컴포넌트에서 `onClick` 이벤트를 `Input` 컴포넌트를 감싸고 있는 `div` 태그에 적용 될 수 있도록 같이 수정하면 좋을 거 같습니다.

aken-you committed:
refactor[#107] : Bottom Sheet의 Backdrop 컴포넌트에 onTap 이벤트 핸들러 추가

aken-you commented:
## 1. 하나의 바텀 시트가 렌더링 된 상태에서, 또 다른 바텀 시트가 렌더링 되는 경우

바텀 시트 라이브러리에서 제공하는 Backdrop 컴포넌트에 onTap 핸들러를 달아서,

```javascript
const SelectBottomSheet = ({ children }: SelectBottomSheetProps) => {
  const { isOpen, onClose } = useSelectContext();

  return (
    <Sheet isOpen={isOpen} onClose={onClose} detent="content-height">
      <Sheet.Container>
        <Sheet.Header />
        <Sheet.Content>{children}</Sheet.Content>
      </Sheet.Container>
      <Sheet.Backdrop onTap={onClose} />
    </Sheet>
  );
};
```
바텀 시트 외부를 클릭할 경우 닫히도록 하였습니다.

https://github.com/user-attachments/assets/03640cff-5035-4caa-8b1e-c7e328b8e1bf

## 2. Select 컴포넌트를 트리거하는 Input

외부에 div로 감싸게 되면, 라벨 옆에 빈 공간을 클릭하게 되더라도 바텀 시트가 열리는 문제가 있습니다.
그래서 만들거면 아예 새로운 컴포넌트로 만드는 게 좋지 않을까 합니다.

https://github.com/user-attachments/assets/5d717bc3-d8e9-4f8e-8034-c0e408f48e74

yonghyeun commented:
### ✅ Test-Storybook Results - Success
```
PASS browser: chromium ../../../../../tmp/6c10e33f343cf3c116bea2964db2d1c4/shared-input.test.js
PASS browser: chromium ../../../../../tmp/6c10e33f343cf3c116bea2964db2d1c4/shared-textarea.test.js (5.002 s)
PASS browser: chromium ../../../../../tmp/6c10e33f343cf3c116bea2964db2d1c4/shared-button.test.js (5.107 s)

Test Suites: 3 passed, 3 total
Tests:       11 passed, 11 total
Snapshots:   0 total
Time:        6.378 s
Ran all test suites.

```
🚀 [스토리북 확인하기](https://66bae48876a5167719ee1778-jqrdxsprac.chromatic.com)
